LaRomance substepping

recuero opened the pull request:
## Reason
Allow robust and error based simulation of reduced-order creep models (i.e. LaRomance)

## Design
Follow the substepping design by @dewenyushu. In general, the creep model will be responsible for managing stateful materials when substepping is used.

## Impact
More robust practical simulations including advanced creep models.

This PR aspires to close #14757.

recuero committed:
Extend substepping implementation to LaRomance (#14757)

recuero committed:
Add tests for LaRomance substepping (#14757)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/891801/) on e2968f6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19414/site)

This comment will be updated on new commits.

dewenyushu commented on modules/tensor_mechanics/src/materials/ADComputeMultipleInelasticStress.C:
```diff
@@ -390,6 +390,9 @@ ADComputeMultipleInelasticStress::computeAdmissibleState(
     ADRankTwoTensor & inelastic_strain_increment)
 {
+  // Properly update material properties (necessary if substepping is employed).
+  _models[model_number]->zeroOutIncrementalMaterialProperties();
```
Do we have the mechanism to prevent calling this if sub-stepping is not enabled for one of the models? 

The users may get into cases when one/several of the multiple inelastic models do not support sub-stepping.

recuero replied:
This would become an empty call up in the hierarchy if the model doesn't support substepping. Calling it should be safe for any model.

dewenyushu commented on modules/tensor_mechanics/include/materials/StressUpdateBase.h:
```diff
@@ -157,6 +157,11 @@ class StressUpdateBaseTempl : public Material
+  /**
+   * Zero out material properties. Useful for substepping with inelastic models.
+   */
+  virtual void zeroOutIncrementalMaterialProperties(){};
```
If this is called, does it mean the user is trying to do sub-stepping but the model does not support it? If so, maybe we should throw an error or something?

recuero replied:
No, you can always call this method. It may or may not be implemented downstream, but that's responsibility of the model. I do think we are not handling an error when the user selects `use_substep = true` and the model doesn't support it (it does not inherit that flag). I'll look into creating that paramError.

dewenyushu commented on modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -90,13 +97,24 @@ RadialReturnStressUpdateTempl<is_ad>::calculateNumberSubsteps(
+    }
+    else
+    {
+      const Real accurate_time_step = _dt * _substep_tolerance / effective_elastic_strain;
+
+      if (_dt > accurate_time_step)
+        number_of_substeps = std::ceil(_dt / accurate_time_step);
```
It seems like this can be refactored so that we do not need `_dt` here:

```
      const Real ratio = _substep_tolerance / effective_elastic_strain;

      if (_ratio<1.0)
        number_of_substeps = std::ceil(1.0 / ratio);
```

dewenyushu commented on modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -277,6 +296,9 @@ RadialReturnStressUpdateTempl<false>::updateStateSubstep(RankTwoTensor & strain_
   const unsigned int total_substeps = calculateNumberSubsteps(strain_increment);
+  _number_of_substeps[_qp] = total_substeps;
+
+  Moose::out << "Number of substeps: " << total_substeps << "\n";
```
This info will be printed our for all `_qp`, right? Maybe this is too much info? Are we going to store this in an additional material property?

recuero replied:
Definitely too much information! That must go!

recuero replied:
Yeah, the one test should be outputting the number of steps to the exodus file.

dewenyushu reviewed (commented):
Really glad that you get this to work. Just a few comments.

lynnmunday commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -856,30 +895,34 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::computeStressFinalize(
-                   ", and/or wall dislocation density, ",
-                   negative_wall_dislocations,
-                   ". Cutting timestep.");
+    mooseError("The negative values of the cell dislocation density, ",
```
This will no longer cut the timestep so you can get rid of that part of that part of the error.

lynnmunday replied:
Maybe this error should be after the verbose output.

recuero replied:
This was a local change I made to make sure I didn't run into these undesirable situations while testing or creating tests (I pushed this as a draft to trigger CIVET testing). I will revert this back to being an exception to avoid any regression in behavior.

recuero committed:
Clean up and address comments (#14757)

The number of substeps material finds a better place in the inelastic model.
In the ComputeMultipleInelasticStress object, there can be various inelastic models
within the same block, so, there,  we can't really tell the number of substeps per
inelastic model per quadrature point.

recuero commented:
@sapitts do we have an example of those numerically challenging reduced order models with large eigenstrains (#14757 - to be used with the existing SS316H model)? It may be good to analyze/include one test with a more meaningful practical setup.

recuero committed:
Catch sneaky situation when user selected substepping but hasn't been implemented.

If this error is not added, ComputeMultipleInelasticStress decides to silently move forward with a
non-substepping algorithm without informing the user. This should allow to independently choose
substepping for inelastic models while erroring out if what the user wants isn't possible.

One test is added to cover this new error.
(#14757)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/893202/) on 7b08065 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>44d8c6</th>
      <th colspan=4>#19414 7b0806</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.83%</td>
      <td>88.86%</td>
      <td>+0.04%</td>
      <td>98.57%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>24793</td>
      <td>24842</td>
      <td>+49</td>
      <td>69</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3118</td>
      <td>3113</td>
      <td>-5</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19414/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19414/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19414/coverage/xfem)

</details>



This comment will be updated on new commits.

dewenyushu commented on modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C:
```diff
@@ -179,6 +179,15 @@ ComputeMultipleInelasticStress::initialSetup()
     paramError("damage_model",
                "Damage Model " + _damage_model->name() +
                    " is not compatible with ComputeMultipleInelasticStress");
+
+  // This check prevents the hierarchy from silently skipping substepping without informing the user
+  for (unsigned int model_number = 0; model_number < _num_models; ++model_number)
+  {
+    const bool use_substep = _models[model_number]->substeppingCapabilityRequested();
+    if (use_substep && _models[model_number]->substeppingCapabilityEnabled() == false)
+      mooseError("Usage of substepping has been requested, but this inelastic model does not "
+                 "implement substepping yet.");
+  }
```
This is great. Thanks @recuero for catching this design defect and making correspondence change.

recuero replied:
Thank you for putting the substepping framework together! I think it's going to see good use in many applied cases.

jiangwen84 commented on modules/tensor_mechanics/src/materials/ADComputeMultipleInelasticStress.C:
```diff
@@ -390,6 +390,9 @@ ADComputeMultipleInelasticStress::computeAdmissibleState(
     ADRankTwoTensor & inelastic_strain_increment)
 {
+  // Properly update material properties (necessary if substepping is employed).
+  _models[model_number]->zeroOutIncrementalMaterialProperties();
```
if it is only necessary when substepping is employed, maybe you could add 
```
if (_use_substep)
```

recuero replied:
This method will provide the value of material properties (which will be zero or "reset") if substepping is not selected. Creep models that include substepping will make use of this method, even if the user has not selected it.

jiangwen84 commented on modules/tensor_mechanics/src/materials/ADPowerLawCreepStressUpdate.C:
```diff
@@ -92,6 +92,18 @@ ADPowerLawCreepStressUpdate::computeStrainEnergyRateDensity(
+void
+ADPowerLawCreepStressUpdate::zeroOutIncrementalMaterialProperties()
+{
+  _creep_strain[_qp] = _creep_strain_old[_qp];
```
same here.

jiangwen84 commented on modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C:
```diff
@@ -466,6 +475,9 @@ ComputeMultipleInelasticStress::computeAdmissibleState(unsigned model_number,
                                                        RankFourTensor & consistent_tangent_operator)
 {
+  // Properly update material properties (necessary if substepping is employed).
+  _models[model_number]->zeroOutIncrementalMaterialProperties();
```
same here.

jiangwen84 commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -156,8 +156,17 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::LAROMANCEStressUpdateBaseTempl(
+        this->_base_name + "wall_dislocations_step")),
+    _cell_dislocations_step(this->template declareGenericProperty<Real, is_ad>(
+        this->_base_name + "cell_dislocations_step")),
+    _plastic_strain_increment(RankTwoTensor::Identity()),
```
why is it initialized as Identify instead of zero?

recuero replied:
Because I was confused about the RankTwoTensor initialization options. I assume `::initNone` is the right option here.

jiangwen84 commented on modules/tensor_mechanics/include/materials/LAROMANCEStressUpdateBase.h:
```diff
@@ -526,6 +530,17 @@ class LAROMANCEStressUpdateBaseTempl : public RadialReturnCreepStressUpdateBaseT
+  RankTwoTensor _plastic_strain_increment;
+
+  /// Material capturing number of substeps. If substepping isn't used will default to one.
+  MaterialProperty<Real> & _number_of_substeps;
```
why is it `Real` type? not `unsigned int`

recuero replied:
Should be `unsigned int` but the materialOutputAction somehow can't handle it. It gives me an error: 

``` *** ERROR ***
test:rom_stress_update.ADrom/lower_limit_extrapolated: The types for total 1 material properties:
test:rom_stress_update.ADrom/lower_limit_extrapolated:   number_of_substeps
test:rom_stress_update.ADrom/lower_limit_extrapolated: are not supported for automatic output by MaterialOutputAction. *** ERROR ***
```
There are very few instances of `MaterialProperty<unsigned int>` in MOOSE.

jiangwen84 replied:
That is wired. My suggestion is either to make `unsigned int` work with `materialOutputAction` or disable material output for the number of substeps using `materialOutputAction`. You can manually output substeps using an auxkernel.

jiangwen84 commented on modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -276,10 +294,10 @@ RadialReturnStressUpdateTempl<false>::updateStateSubstep(RankTwoTensor & strain_
                                                          RankFourTensor & tangent_operator)
 {
-  const unsigned int total_substeps = calculateNumberSubsteps(strain_increment);
+  const unsigned int total_number_steps = calculateNumberSubsteps(strain_increment);
```
to be consistent, `total_number_substeps` would be better?

jiangwen84 commented on modules/tensor_mechanics/src/materials/StressUpdateBase.C:
```diff
@@ -109,7 +109,7 @@ TangentCalculationMethod
 {
   mooseError(
-      "getTangentCalculationMethod called: no tangent calculationg is needed while using AD");
+      "getTangentCalculationMethod called: no tangent matrix calculation is needed while using AD");
```
suggestion: no tangent **moduli** calculation

jiangwen84 commented on modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -90,13 +97,24 @@ RadialReturnStressUpdateTempl<is_ad>::calculateNumberSubsteps(
+      const Real ratio = effective_elastic_strain / _max_inelastic_increment;
+
+      if (ratio > _substep_tolerance)
+        number_of_substeps = std::ceil(ratio / _substep_tolerance);
```
We probably need to set up an upper limit for the number_of_substeps too. According to my experience with crystal plasticity, when number of sub stepping is greater than 10, it no longer helps the convergence and would slow down the calculation. I do know what is a best number to use here though. :-)

recuero replied:
This is a great point. @dewenyushu and I discussed this. It'd be good to know the reason why increased number of substeps doesn't help in some of these cases. I'd think that BCs and other loading may play a role in this, i.e. it may depend on the simulation setup. We can create a parameter that the user can override and throw an exception if it turns out the number is larger? I will include this change and see if there are any failures.

recuero replied:
A user may want to achieve a particular numerical creep integration error at the expense of more substeps... We should probably be generous with the default number of substeps. I can try 25 initially.

jiangwen84 reviewed (dismissed):
Looks great. Several minor comments.

recuero commented on modules/tensor_mechanics/src/materials/PowerLawCreepStressUpdate.C:
```diff
@@ -94,6 +94,18 @@ PowerLawCreepStressUpdate::computeStrainEnergyRateDensity(
+void
+PowerLawCreepStressUpdate::zeroOutIncrementalMaterialProperties()
+{
+  _creep_strain[_qp] = _creep_strain_old[_qp];
```
This method can do different things in different material models. The idea is to provide the creep model the ability of including substepping in its logic. LaRomance sets incremental properties to zero, whereas this method resets `creep_strain`, thereby zeroing out the existing incremental values. I will replace the name of this method with `resetIncrementalMaterialProperties`.

recuero committed:
Address Wen's review (#14757)

recuero committed:
Add doco and mooseDoOnce for LaRomance valid range warning (#14757)

This warning can fully populate MOOSE's output. This can be detrimental
to clean iteration output, which is relevant for more applied runs.
This warning is not to be dismissed since it's informing the user of
exercising a reduced order model outside the training range, but we here
assume that printing it to the screen once would alert the user of this
fact.

recuero commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -489,7 +528,7 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::checkInputWindow(const GenericReal<is_ad>
     {
       case WindowFailure::WARN:
-        mooseWarning(msg.str());
+        mooseDoOnce(mooseWarning(msg.str()));
```
@tophmatthews are you okay with this change? We were thinking the warning output can become too invasive when we evaluate the model outside the range and that, hopefully, printing it out once would be enough to alert the user. Let me know if you think otherwise and I'll revert it.

tophmatthews replied:
Isn't mooseWarning already wrapped in a DoOnce?

recuero replied:
`mooseDeprecatedStream` and `mooseInfoStream` are, but not `mooseWarning`.

tophmatthews replied:
Yeah, this is probably fine. We're working hard on removing the need for this, so seems good to me.

recuero commented:
@jiangwen84 thanks for the review. @lynnmunday volunteered to give this a try on some cool models he prepared for a project. So, whenever you are okay with the PR, it can be merged.

recuero commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -156,8 +156,17 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::LAROMANCEStressUpdateBaseTempl(
+    _cell_dislocations_step(this->template declareGenericProperty<Real, is_ad>(
+        this->_base_name + "cell_dislocations_step")),
+    _plastic_strain_increment(RankTwoTensor::Identity()),
+    _number_of_substeps(
+        this->template declareProperty<Real>(this->_base_name + "number_of_substeps"))
```
See https://github.com/idaholab/moose/pull/19414#discussion_r754470870.

tophmatthews commented:
Just for my own curiosity, the initial capability was setup by @dewenyushu a year or so ago, and this is the implementation of that capability for creep models?

tophmatthews commented:
Super stoked to see the implementation here! Certainly for metallic fuel problems, the max inelastic timestep really kills our wall time by requiring many small timesteps.

recuero commented:
There are a number of design decisions to be made, for auxiliary kernels. The existing ones are already templated with AD types, which are not compatible with `Int`. A hierarchy parallel to `MaterialAuxBase` appears necessary (not just creating `MaterialUnsignedIntAux`)

dschwen commented:
My suggestion would be to create a universal `MaterialIntegerAux` AuxKernel that tries all integer flavors (using `haveMaterialProperty<>()` (e.g. `int`, `unsigned int`, `char`, `short`, `long`, `long long`, `unsigned short`, etc.)

recuero commented:
Anyway, this is all quite of an excursion from the purpose of this PR. I'll open a separate one when I get to it.

lynnmunday commented:
Can you merge this without fixing the unsigned int material output and then fix it later?

jiangwen84 commented:
> Can you merge this without fixing the unsigned int material output and then fix it later?

I think @recuero can use `unsigned int` without outputting them. Another PR to add the material output capability for `unsigned int`. Any thoughts?

recuero commented:
>can use `unsigned int` without outputting them

Changes are required to the MaterialsOutputAction to use `unsigned int` in the material property. Even tests that were not explicitly outputting the variable were failing with the same message. As this PR stands right now we are casting an integer into a real for output purposes. AFAICS this should not cause problems other than a pesky "x.00" in the output maybe.

I agree with @lynnmunday that the use and testing of the capability in practical problems should have priority in this case.

jiangwen84 commented:
> > can use `unsigned int` without outputting them
> 
> Changes are required to the MaterialsOutputAction to use `unsigned int` in the material property. Even tests that were not explicitly outputting the variable were failing with the same message. As this PR stands right now we are casting an integer into a real for output purposes. AFAICS this should not cause problems other than a pesky "x.00" in the output maybe.
> 
> I agree with @lynnmunday that the use and testing of the capability in practical problems should have priority in this case.

Ok. I am fine with using `Real`. I will leave @dschwen for the final merge. 

BTW, @lynnmunday can pull down your PR and test it locally without waiting this being merged.

dschwen commented on modules/tensor_mechanics/doc/content/source/materials/RadialReturnStressUpdate.md:
```diff
@@ -75,6 +75,41 @@ Users can print out any of these strains and stresses using the `RankTwoAux` as
+
+Regular use of the radial return mapping triggers one instance of the return to the yield surface that spans
+the entire system wide time step. While the material time step limiter discussed in [ComputeMultipleInelasticStress](ComputeMultipleInelasticStress.md) can
+effectively limits the time step size to achieve the desired convergence or integration error properties, it does so at
```
```suggestion
effectively limit the time step size to achieve the desired convergence or integration error properties, it does so at
```

dschwen commented on modules/tensor_mechanics/include/materials/LAROMANCEStressUpdateBase.h:
```diff
@@ -526,6 +530,17 @@ class LAROMANCEStressUpdateBaseTempl : public RadialReturnCreepStressUpdateBaseT
+  /// Total plastic strain increment in step (summing substep contributions)
+  RankTwoTensor _plastic_strain_increment;
+
+  /// Material capturing number of substeps. If substepping isn't used will default to one.
```
```suggestion
  /// Material property capturing number of substeps for output purposes (defaults to one if substepping isn't used) ```

recuero replied:
Check

dschwen commented on modules/tensor_mechanics/include/materials/StressUpdateBase.h:
```diff
@@ -141,6 +141,13 @@ class StressUpdateBaseTempl : public Material
+  /**
+   * Has the user requested usage of (possibly) implemented substepping capability for inelastic
+   * models.
+   *
```
```suggestion
```

recuero replied:
Done

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C:
```diff
@@ -179,6 +179,15 @@ ComputeMultipleInelasticStress::initialSetup()
                    " is not compatible with ComputeMultipleInelasticStress");
+
+  // This check prevents the hierarchy from silently skipping substepping without informing the user
+  for (unsigned int model_number = 0; model_number < _num_models; ++model_number)
```
Just a _suggestion_ (automatically uses the correct type for the counter - requires `#include "libmesh/int_range.h"`).

```suggestion
  for (auto model_number: index_range(_models))
```

recuero replied:
I just _made_ that change.

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C:
```diff
@@ -179,6 +179,15 @@ ComputeMultipleInelasticStress::initialSetup()
+  {
+    const bool use_substep = _models[model_number]->substeppingCapabilityRequested();
+    if (use_substep && _models[model_number]->substeppingCapabilityEnabled() == false)
+      mooseError("Usage of substepping has been requested, but this inelastic model does not "
```
How will the user know what "this inelastic model" is if multiple models are used?

recuero replied:
Should be addressed now.

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C:
```diff
@@ -466,6 +475,9 @@ ComputeMultipleInelasticStress::computeAdmissibleState(unsigned model_number,
                                                        RankTwoTensor & inelastic_strain_increment,
                                                        RankFourTensor & consistent_tangent_operator)
 {
+  // Properly update material properties (necessary if substepping is employed).
```
"properly update" is not very descriptive. Is it maybe "reset deformation state to the beginning of the increment" or something along those lines?

recuero replied:
Yeah, it's resetting material properties.

dschwen commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -156,8 +156,17 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::LAROMANCEStressUpdateBaseTempl(
+        this->_base_name + "wall_dislocations_step")),
+    _cell_dislocations_step(this->template declareGenericProperty<Real, is_ad>(
+        this->_base_name + "cell_dislocations_step")),
+    _plastic_strain_increment(RankTwoTensor::initNone),
```
Why `initNone` if you zero it out below? You could just do 
```suggestion
    _plastic_strain_increment(),
```

recuero replied:
Ok

dschwen commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -349,6 +358,30 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::maximumPermissibleValue(
+void
+LAROMANCEStressUpdateBaseTempl<is_ad>::resetIncrementalMaterialProperties()
+{
+
```
```suggestion
```

recuero committed:
Address Daniel's review (#14757)

dschwen commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -161,7 +161,7 @@ LAROMANCEStressUpdateBaseTempl<is_ad>::LAROMANCEStressUpdateBaseTempl(
     _cell_dislocations_step(this->template declareGenericProperty<Real, is_ad>(
         this->_base_name + "cell_dislocations_step")),
-    _plastic_strain_increment(RankTwoTensor::initNone),
+    _plastic_strain_increment(),
```
You don't need to zero below
